src/auditwheel/mod.rs

```diff
@@ -0,0 +1,249 @@
+mod policy;
+
+use crate::Manylinux;
+use crate::Target;
+use anyhow::Result;
+use fs_err::File;
+use goblin::elf::{sym::STT_FUNC, Elf};
+use goblin::strtab::Strtab;
+use regex::Regex;
+use scroll::Pread;
+use std::collections::{HashMap, HashSet};
+use std::io;
+use std::io::Read;
+use std::path::Path;
+use thiserror::Error;
+
+use policy::POLICIES;
+
+/// Error raised during auditing an elf file for manylinux compatibility
+#[derive(Error, Debug)]
+#[error("Ensuring manylinux compliance failed")]
+pub enum AuditWheelError {
+    /// The wheel couldn't be read
+    #[error("Failed to read the wheel")]
+    IOError(#[source] io::Error),
+    /// Reexports elfkit parsing errors
+    #[error("Goblin failed to parse the elf file")]
+    GoblinError(#[source] goblin::error::Error),
+    /// The elf file isn't manylinux compatible. Contains the list of offending
+    /// libraries.
+    #[error(
+    "Your library links libpython ({0}), which libraries must not do. Have you forgotten to activate the extension-module feature?",
+    )]
+    LinksLibPythonError(String),
+    /// The elf file isn't manylinux compatible. Contains the list of offending
+    /// libraries.
+    #[error(
+    "Your library is not manylinux compliant because it links the following forbidden libraries: {0:?}",
+    )]
+    ManylinuxValidationError(Vec<String>),
+}
+
+/// Structure of "version needed" entries is documented in
+/// https://refspecs.linuxfoundation.org/LSB_3.0.0/LSB-PDA/LSB-PDA.junk/symversion.html
+#[derive(Clone, Copy, Debug, Pread)]
+#[repr(C)]
+struct GnuVersionNeed {
```

Comment:
Could those two structs (and maybe that relevant reading part) be upstreamed to goblin?

Reply:
I'd love to, but I'll do it separately to not block this.